### ISSUE TITLE
dnsmasq fails to identify LAN requests when started before network

### ISSUE THREAD

**Issue by MichaIng** [pi-hole/pi-hole#2924]
**In raising this issue, I confirm the following:**

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*.
- [x] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:** `9`

---
**Expected behaviour:**

When using `pihole -a -i local`, after reboot, Pi-hole is expected to allow DNS requests from local network clients.

**Actual behaviour:**

Local network requests are denied by the underlying `dnsmasq`, producing the log entry:
```
Sep 14 16:36:16 dnsmasq[301]: Ignoring query from non-local network
```

**Steps to reproduce:**

1. `pihole -a -i local`
2. `reboot`
3. E.g. `wget --spider pi-hole.net` from a client, located in the same local network.

**Troubleshooting undertaken, and/or other relevant information:**

Bug report, investigation and identifying the underlying issue in dnsmasq has been done here: https://github.com/MichaIng/DietPi/issues/3007#issuecomment-522942038
- `pihole -a -i local` adds the dnsmasq setting: [`local-service`](https://manpages.debian.org/buster/dnsmasq-base-lua/dnsmasq.8.en.html)
- When `dnsmasq` starts with this setting before any network interfaces has been configured, it does not have any network or IP info where it could derive the information from, which client IPs are from within the local network and which not, which is logically understandable. Obviously a DNS request does not force `dnsmasq` to update this info.

The following can be done to fix the issue:
- Manually restart `pihole-FTL`.
- Add startup dependencies to service to have it started after network interfaces being configured: `# Required-Start: $network` => `After=network-online.target`
- Access the Pi-hole admin panel (web UI), even without login, although I did not yet identify the exact trigger from e.g. the php script.

**Suggested solution:**
- Add network to the start requirement of the pihole-FTL service.
```
# Required-Start:    $remote_fs $syslog $network
# Required-Stop:     $remote_fs $syslog $network
```
- If systemd is used, this leads to [`After=network-online.target`](https://www.freedesktop.org/software/systemd/man/systemd.special.html#network-online.target) being added to the generated systemd unit.
- It generally makes sense, not only to prevent this issue, but also since no DNS resolution can be done without network access anyway.
- Additionally this can be taken as initiator to create a proper `pihole-FTL.service` systemd unit instead of relying on the sysvinit => systemd generator on very most modern distros, which is not assured to translate everything very well. E.g. `$syslog` seems to be not translated at all on Debian, even with `rsyslog` installed additionally to `systemd-journald`:
```
root@VM-Buster:~# systemctl cat pihole-FTL.service
# /run/systemd/generator.late/pihole-FTL.service
# Automatically generated by systemd-sysv-generator

[Unit]
Documentation=man:systemd-sysv-generator(8)
SourcePath=/etc/init.d/pihole-FTL
Description=LSB: pihole-FTL daemon
Before=multi-user.target
Before=multi-user.target
Before=multi-user.target
Before=graphical.target
After=remote-fs.target

[Service]
Type=forking
Restart=no
TimeoutSec=5min
IgnoreSIGPIPE=no
KillMode=process
GuessMainPID=no
RemainAfterExit=yes
SuccessExitStatus=5 6
ExecStart=/etc/init.d/pihole-FTL start
ExecStop=/etc/init.d/pihole-FTL stop
ExecReload=/etc/init.d/pihole-FTL reload
```

**Comment by dschaper** [comment]
See https://github.com/pi-hole/pi-hole/pull/2900 for already existing discussion and proposed service file.

**Comment by MichaIng** [comment]
@dschaper @DL6ER 
I just recognised this open issue again and remembered: https://discourse.pi-hole.net/t/have-to-pihole-restartdns-after-reboot/28772/29?u=michaing
So basically as I suggested above already, adding `$network` to the service would at least reduce the risk for any such related issues. What do you think?

I'll open a PR.

**Comment by dschaper** [comment]
Will it work for the many instances that don't use systemd?

**Comment by MichaIng** [comment]
PR up: https://github.com/pi-hole/pi-hole/pull/3248

@dschaper 
To be true I am not sure how service ordering works on sysvinit, but `$network` is a sysvinit "syntax" that is only interpreted by systemd for backwards compatibility. On sysvinit I can only assume it has the same effect, delaying the service start until network is up, why would it otherwise exist? But I'll do some research. At least we should make sure that it cannot cause unwanted service hangs.

**Comment by MichaIng** [comment]
Ref: http://refspecs.linuxfoundation.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/facilname.html
> $network  basic networking support is available. Example: a server program could listen on a socket.

**Comment by pralor** [comment]
This issue has been mentioned on **Pi-hole Userspace**. There might be relevant details there:

https://discourse.pi-hole.net/t/pihole-not-listening-on-wireguard-interface-despite-listen-on-all-interfaces-is-set/30212/2

**Comment by MichaIng** [comment]
PR has been merged for Pi-hole v5.1, I mark this as closed.
